### DISCUSSION TITLE
Pause/Interrupt Long running Tasks or Split in several Jobs

### DISCUSSION THREAD

**Discussion by svenbaus** [quartznet/quartznet#1284] [Q&A] [answered]
Hello together,

We are planning to use Quartz Scheduler in a Enterprise Environment Blazor Server Application which will be a distributed system. Most part will run on Azure, but some parts need to run on a internal worker because of security reasons.

As I am currently designing the system architektural I have a question for the best approach. The basic approach will be a database queue to the internal worker, which will grab internal resources. The Azure side will push requests in this database queue and the internal worker will actually work this requests. On the Azure side we have IJob instances, that process data requested by the user. Inside some of this IJob instances, we might need internal resources.

So my question is:
1. What is the difference between pausing and interrupting a IJob?
2. Is it a good idea to pause the IJob instance on the azure side when waiting for internal resources or would it be better to have a cascade of IJob instances (for example one to gather internal resources, one to process data for user)?

Thanks for your comments and thanks for this framework :).

**Comment by lahma** [comment, marked as answer]
By the sounds of it, I would question whether scheduling library is the right choice here (even if being called a bad salesman for that). If you have a *queue* that you want to process - why don't you want to use a messaging solution like ASB? That paired with something like MassTransit might fit the bill.  You could have internal workers (consumers) connecting to public ASB queue I guess.

But regardless, about the questions you asked.

> What is the difference between pausing and interrupting a IJob?

Pausing a job prevents it from processing anything that would be processed next based on trigger schedule. Triggers associated with the job won't fire as job is paused - same goes with pausing a trigger (jobs are run by triggers that are not paused) and pausing a scheduler (nothing gets run).

Interrupting a job means signaling a *running* job that it should terminate processing, this will require job the observe the cancellation token in context, so a job which doesn't observe the token won't interrupt properly.

> Is it a good idea to pause the IJob instance on the azure side when waiting for internal resources or would it be better to have a cascade of IJob instances (for example one to gather internal resources, one to process data for user)?

I probably don't quite understand the question. I would create a job per "use case". You can also configure whether job can run concurrently (`DisallowConcurrentExceutionAttribute`) or limit active trigger execution count (thread pool size). But again sounds more like a queueing problem than scheduling problem.

Hope this helps.

**Reply by svenbaus** [reply]
Thanks for your very good answer :). This clears the difference for pausing and interrupting a job.

Since I'm currently not that familiar with queues like IBM or AzureServiceBus (you mean this by ASB, correct?) I took the idea of having a queue in form of a database, but I'm currently not happy with this (concurrent access of 2 systems can be very bad). So I will have a look into queues and especially ASB.
In the Blazor Application hosted on Azure we will use Quartz and maybe have the requests to internal resources in a queue.

With your answer I already thought of having several Jobs running in order to meet our needs. Something like 1. Start processing of userdata -> 2.request internal resources -> 3. Process internal resources -> 4. Process userdata together with internal resources and present the result.

Thanks for your help :)!

**Reply by lahma** [reply]
Yes, by ASB I meant the Azure service, like you deducted. Glad that I was able to help!
